lib/types.nix

```diff
@@ -302,6 +302,35 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+
+    orderOf = {
+      before ? a: b: false,
+      elemType
+    }: mkOptionType rec {
+      name = "orderOf";
+      description = "ordered entries of ${elemType.description}s";
+      check = elemType.check;
+      merge = loc: defs:
+        let
+          nodeAttributes = (attrsOf elemType).merge loc defs;
+          entries = mapAttrsToList nameValuePair nodeAttributes;
+          sortedEntries = toposort before entries;
```

Comment:
`toposort` is `O(n^2)`, so I’m kinda afraid this might slow down the module system even more if it’s used in a few places.

Reply:
I'm not sure if it's possible to implement `toposort` more efficiently. However I can look into whether it's possible to implement a different interface more efficiently, e.g. something that isn't based on a `before` function but on a successor mapping instead, e.g. `graphToposort { foo = [ "bar" ]; bar = []; }` representing `foo` having a directed edge to `bar`. In theory, topo sorts can be `O(|N| + |E|)`, though that might not be implementable in Nix.